scripts/train/train.py

```diff
@@ -181,153 +200,232 @@ def main(cfg: DictConfig):
 
-    # Run Name
-    if cfg.get('run_name') is None:
-        cfg.run_name = os.environ.get('RUN_NAME', 'llm')
+    # Initialize distributed setting with seed
+    dist_timeout: int = cfg.pop('dist_timeout', 600.0)
+    dist.initialize_dist(get_device(None), timeout=dist_timeout)
 
-    # Get batch size info
+    # Get global and device batch size information from distributed/single node setting
+    # TODO remove this as we should treat configs as immutable objects
     cfg = update_batch_size_info(cfg)
 
-    # Read FSDP Config as a dict
-    fsdp_config = cfg.get('fsdp_config', None)
-    fsdp_config = om.to_container(fsdp_config,
-                                  resolve=True) if fsdp_config else None
-    assert isinstance(fsdp_config, Dict) or fsdp_config is None
+    print('Logging config...')
+    log_config(cfg)
+
+    # Mandatory model training configs
+    model_config: DictConfig = pop_config(cfg, 'model', must_exist=True)
+    tokenizer_config: DictConfig = pop_config(cfg, 'tokenizer', must_exist=True)
+    optimizer_config: DictConfig = pop_config(cfg, 'optimizer', must_exist=True)
+    scheduler_config: DictConfig = pop_config(cfg, 'scheduler', must_exist=True)
+    train_loader_config: DictConfig = pop_config(cfg,
+                                                 'train_loader',
+                                                 must_exist=True)
+
+    # Optional fsdp data, fine-tuning, and eval configs
+    fsdp_config: Optional[DictConfig] = pop_config(cfg,
+                                                   'fsdp_config',
+                                                   must_exist=False)
+    lora_config: Optional[DictConfig] = pop_config(cfg,
+                                                   'lora',
+                                                   must_exist=False)
+    eval_loader_config: Optional[DictConfig] = pop_config(cfg,
+                                                          'eval_loader',
+                                                          must_exist=False)
+    icl_tasks_config: Optional[DictConfig] = pop_config(cfg,
+                                                        'icl_tasks',
+                                                        must_exist=False)
+
+    # Optional logging, evaluation and callback configs
+    logger_configs: Optional[DictConfig] = pop_config(cfg,
+                                                      'loggers',
+                                                      must_exist=False)
+    callback_configs: Optional[DictConfig] = pop_config(cfg,
+                                                        'callbacks',
+                                                        must_exist=False)
+    algorithm_configs: Optional[DictConfig] = pop_config(cfg,
+                                                         'algorithms',
+                                                         must_exist=False)
+
+    device_train_batch_size: int = cfg.pop('device_train_batch_size')
```

Comment:
Should we:
- have type annotations? 
- use `pop_config` function in which case the warnings will be very verbose which may be good? 
- list all the DEFAULT_VALUES at the top of the file?